### ISSUE TITLE
@overload cannot replace previously defined or built-in implementations

### ISSUE THREAD

**Issue by ARF1** [numba/numba#5043]
Let's assume I would like to `@overload` (or replace) an already existing implementation of a numpy function. In this case, the `@overload` function seems to get called (i.e. it produces errors e.g. if the returned function signature does not match) but the new implementation never gets used:

(Tested with numba 0.47)

```
import numpy as np
from numba import  njit
import numba.types as types
from numba.extending import overload, overload_method, register_jitable

@overload(np.sum)
@overload_method(types.Array, "sum")
def array_sum(arr):
    def array_sum_impl(arr):
        print('Test1')
        return 0.0
    return array_sum_impl

@njit
def test_sum(arr):
    return np.sum(arr)

arr = np.arange(10)
print(test_sum(arr))
```

Prints `45.0` rather than `0.0`.

----
Similarly for previously defined implementations that are not built-in:

```
import numpy as np
from numba import  njit
import numba.types as types
from numba.extending import overload, overload_method, register_jitable

@overload(np.apply_along_axis)
def array_sum(arr):
    def array_sum_impl(arr):
        return 0.0
    return array_sum_impl

@njit
def test_sum(arr):
    return np.apply_along_axis(arr)

arr = np.arange(10)
print(test_sum(arr))

@overload(np.apply_along_axis)
def array_sum(arr):
    def array_sum_impl(arr):
        return 1.0
    return array_sum_impl

@njit
def test_sum(arr):
    return np.apply_along_axis(arr)
print(test_sum(arr))
```

Prints
```
0.0
0.0
```

rather than 
```
0.0
1.0
```

**Comment by stuartarchibald** [comment]
Thanks for the report. The bevahiour herein is expected given the implementation, will discuss with the other core developers.

**Comment by ARF1** [comment]
@stuartarchibald Thanks for responding. I half-expected that it was intended behaviour. I was not sure though because I could not find any documentation of this (to me) unexpected behaviour. It took me ages to figure out, why my implementation did not seem to work as expected.

If you decide to keep the current behaviour, would it be feasible to print a warning when a user attempts to redefine and already defined implementation? I believe that would be really helpful.

**Comment by stuartarchibald** [comment]
> @stuartarchibald Thanks for responding. I half-expected that it was intended behaviour. I was not sure though because I could not find any documentation of this (to me) unexpected behaviour. It took me ages to figure out, why my implementation did not seem to work as expected.
> 
> If you decide to keep the current behaviour, would it be feasible to print a warning when a user attempts to redefine and already defined implementation? I believe that would be really helpful.

Yes, we're about to undertake a load of refactoring of internals to move entirely to use the extension API, I think it'd be sensible to make redefinition a hard error as it's probably a mistake/will not have the desired effect as noted.

**Comment by ARF1** [comment]
Thanks. An error would be a good solution.

That said (and having no idea of the implementation difficulty), I would like point out that there are two arguments for making redefinitions work:

- for rapid prototyping it would be very convenient if `@overload` would allow redefinitions. Also, allowing redefinitions is kind of a python convention...
- allowing usert to extend existing implementations without having to recompile numba in its entirety. (See example below.)
----
_Example for extending existing implementations:_

Take `numpy.argsort()`: Numba has support for the keyword parameter `kind` with values `quicksort` and `mergesort`. The user however has code that uses `heapsort`.

It would very convenient for users to be able to use python patterns they are used to from monkey-patching to overwrite the `argsort` implementation with one that handles the `kind` value of `heapsort` and delegates all other input options to the existing implementation:

Something along the lines of:
```python

previous_argsort_impl = get_existing_implementation(np.argsort, types.Array, ...)

@overload(np.argsort, types.Array, ...)
def array_argsort_impl(arr, kind='quicksort'):
    if kind == 'heapsort':
        def new_implementation(...):
            #code
        return new_implementation
    else:
        return previous_argsort_impl
```

With prohibited redefinitions though, the user will have to recompile numba which arguably is a much higher "barrier to entry" for those users that are primarily data scientists and developers a distant second.

**Comment by stuartarchibald** [comment]
I can see some use for the redefinition behaviour but also feel that if it's a missing piece of functionality in Numba it should really just go into Numba, e.g. the `heapsort` example above.

**Comment by ARF1** [comment]
> I can see some use for the redefinition behaviour but also feel that if it's a missing piece of functionality in Numba it should really just go into Numba, e.g. the `heapsort` example above.

I agree that missing functionality should go into numba.

Though normal users generally have a specific problem they want to solve **right now**. I think they are more likely to write an implementation for missing functionality if they can get started without:
- cloning numba
- delving too deep into numba internals to find where their function is actually implemented
- trying to understand how the implementation is structured and how they can fit their extension into it
- rebuilding numba
- thinking about how they can get their newly built numba onto their production machines
- etc...

They just want to make their code compile... From experience, I can tell you that the learning curve is steep in the beginning. - Possibly too steep if somebody just wants to get their work done...

With redefinitions, they can skip all the painful steps above, solve their problem and get on with their day.

Once people have written an implementation, they might be inclined to clean it up and submit it as well. Many will probably be proud that they actually managed to get it working and want to share their success with the world. ;-)

But in the end this is a policy decision. Thanks for taking the time to consider it.

For what it is worth, I have no real objection to making redefinitions a hard error.

**Comment by pearu** [comment]
Making a redefinition a hard-error can be reasonable when redefinition is done by accident.
However, if the redefinition can be supported in Numba such that the redefinition act itself is solid and well-defined then it should be supported, IMHO. My reasoning is as follows.

Allowing redefinitions gives a developer more freedom to experiment with the software. This increases creativity. Notice that in Python one can redefine even builtin functions or functions from external packages without problems. In fact, this behavior in Python supports solving real problems at hand. 

Note that such user extensions from redefining the builtin implementations do not necessarily mean that the upstream software is lacking the required functionality. At the point of working with problems, it is not clear if the extensions should be made available to upstream or not. In either case, allowing redefinitions allows working with the problems.

However, disallowing extensions (by making redefinitions hard-errors, for instance) means that all the above is made harder for developers and actually it will slim the chances of upstream getting PRs that would extend the existing functionality. Although this might be convenient for the maintainers and good for gaining stability, it is also a seed for stagnation.

**Comment by stuartarchibald** [comment]
@pearu I think redefinition can be useful, for similar reasons to those noted, but it is a challenging thing to support. I'm also of the view that Numba is not in the position right now to handle redefinition and all the complexity that it brings. Given the current code base does poorly defined/unexpected/relying on implementation details type things, I would rather just prevent problems until a solution is well defined. The "Hard Error" behaviour is not immutable, some future release that does a good job of handling redefinition semantics can have it removed :)

**Comment by dlee992** [comment]
This could be a nice feature to have. I will try to give a hacky PR later. Any clue for me to implement this before I really do this? cc @stuartarchibald

**Comment by dlee992** [comment]
I thought before we have the real feature, here is a workaround, Numba can compile it and choose the right one at runtime.
```python
import numpy as np
from numba import njit, types
from numba.extending import overload
from numba.typed import Dict

def foo(a, b, c):
    pass

@overload(foo)
def ol_xx(a, b, c):
    if isinstance(c, types.Integer):
        def impl(a, b, c):
            a = a + b
            return a
        return impl
    if isinstance(c, types.Boolean):
        def impl(a, b, c):
            a = a - b
            return a
        return impl
    if isinstance(c, types.Float):
        def impl(a, b, c):
            a[1] = a[1] + b
            return a
        return impl

@njit(cache=True)
def goo(a, b, c):
    return foo(a, b, c)

arr = np.array([1, 2, 3])
dct = Dict.empty(types.int64, types.int64)
dct[1] = 2
print(goo(arr, 2, 1))
print(goo(1, 2, True))
print(goo(dct, 2, 3.0))
```

**Comment by dlee992** [comment]
with this PR, if add `@overload(np.apply_along_axis, override=True)` for the second definition, the output is correct.

**Comment by gmarkall** [comment]
> with this PR, if add `@overload(np.apply_along_axis, override=True)` for the second definition, the output is correct.

#9578, right?

**Comment by dlee992** [comment]
Yeah, it's more like a POC. But I would like to push it forward to make it complete. But I can enjoy a "pre" review if anyone would like to comment.
